### ISSUE TITLE
GNS: lopsided description of namespace consistency

### ISSUE THREAD

**Issue by peterthomassen** [nweaver-blockchain/blockchain-names#3]
I find some of the description of namespace consistency in GNS rather questionable, as the DNS suffers from the same observed aspects, but the text reads as if these characteristics were weak design on part of the GNS. Let me explain.

From the text:
> Instead, users are intended to create a local configuration mapping human readable zone names to public keys, creating a deliberately inconsistent global namespace.

If the alleged inconsistency comes from the fact that configuration is necessary, and different configuration might lead to different namespaces, the following about the DNS is also correct:

"For the DNS, resolver operators are intended to create a local configuration mapping the root domain onto the root zone nameserver IP addresses, *creating a deliberately inconsistent namespace*."

Now, I feel that the last part of this sentence feels very wrong. Of course, you can configure a resolver to use a different root -- but that's not the same as saying that the DNS creates "a deliberately inconsistent namespace". I think the GNS-related text reads weird in the same way.

---

From the text:
> However, this approach results in the top level domains of the global namespace only consisting of non-human-readable entities, such as example.000G006K2TJNMD9VTCYRX7BRVV3HAEPS15E6NHDXKPJA1KAJJEG9AFF884. A GNS client however can internally specify “Pet Names”, a mapping of a DNS style namespace to a particular zTLD, so .gnu might refer to 000G006K2TJNMD9VTCYRX7BRVV3HAEPS15E6NHDXKPJA1KAJJEG9AFF884, so example.gnu could be resolved by a GNS client.
> 
> The result is that although the global cryptographic namespace is consistent among all clients, any fully human readable names will be inconsistent unless all users use a common source for their pet names.

This mechanism essentially ties a public key to a name. The same is true for the DNS: validators need to configure a trust anchor for the root zone, which is not human-readable. Like in GNS, DNSSEC trust anchors can be configured for any name (not just the root, but also any other names down the tree).

So it seems to me that the degree to which naming consistency is achieved depends on configuration (and access to a shared source) in pretty much the same way. Thus:

"For the DNS, the result is any fully human readable names will be inconsistent unless all users use a common source for their trust anchors."

In fact, the text acknowledges that "the global cryptographic namespace is consistent among all clients", so GNS provides higher consistency than DNS (namely, at least in the cryptographic namespace; the DNS however provides consistency a priori in no namespace).

---

In the DNS world, consistency problems are resolved by getting everyone to configure the de-facto one-and-only IANA root zone, with their trust anchors. And that's fine. But, a very similar de-facto standard is possible with GNS. The text says:

> The largest concern with GNS is the lack of a globally consistent human-readable namespace. It is impossible to ensure that pet-names are consistent, thus all GNS records should be identified by zTLD.

What is the basis for the claim that ensuring consistency is impossible? It's certainly not impossible for the DNS (although the system itself does not provide consistency), so why would it be different for GNS?

(Note: I think it's very unlikely that GNS will attain a sufficient user base to become a serious player, and as a result I'm by no means a GNS proponent. I instead spend a lot of time on making DNSSEC easier to use. But having read the GNS spec, I felt like this text could benefit significantly from a more balanced description.)

**Comment by nweaver** [comment]
Ack.  Added language that the introduction of a single central authority colud be used to create a human consistent namespace, but that runs counter to the goals of decentralization and censorship resistance.

**Comment by peterthomassen** [comment]
Thanks. I think the conclusion now needs adjustment for consistency -- the added sentence said that a consistent human-readable namespace is possible, while the conclusion says it isn't.

I also don't think that this is right:
> users are intended to create a local configuration mapping human readable zone names to public keys, creating a deliberately inconsistent global namespace.

The first part is right, but the second isn't. It's entirely feasible for everyone to agree on some common configuration, and in fact I think that *if* GNS was used broadly, that would be the case. (As discussed, same goes for DNS.) The text, however, sounds like it isn't really working.

Note also that GNS provides all guarantees that DNS provides (including those of DNSSEC), but it has some added privacy guarantees. From a theoretical position, it is thus an alternative, perhaps even better system. I think that should be called out. (And it should also be called out that a transition is pretty much not imaginable.)

Lastly, why is this labeled a "Blockchain (and Blockchain-adjacent) Naming System" on the landing page? As far as I can see, it has nothing to do with it, and its storage concept is more like that of DNS (including no generally copy-distributed public storage).

**Comment by nweaver** [comment]
GNS evolved out of very much the same "decentralized" blockchain ethos and gets specific support from various "blockchain" institutions.  

Would Blockchain (and other 'decentralized') naming sound better?

**Comment by nweaver** [comment]
I'll do a bit more elaboration on the privacy aspects, it does lose availability however.

**Comment by nweaver** [comment]
(note it will probably need to wait for a couple of days however, I've got to head to an unrelated meeting).

**Comment by peterthomassen** [comment]
> Would Blockchain (and other 'decentralized') naming sound better?

Very much, yes, although why the quotes around "decentralized"?

I think decentralization (with or without ethos, either way) is a good goal (the DNS is also decentralized). Blockchain, OTOH, typically is "solution solving" and usually (IMHO) not the way to go. (It's pretty clear from your analysis that e.g. the costs and resource demands of ENS are quite prohibitive, and I agree with all of that.)

I didn't know that GNS arose out of blockchain-related circles. In fact, I can't find anything related to that (but I'd love to learn more about it). For example, the [announcement of the first specification milestone](https://www.gnunet.org/en/news/2019-10-GNSSpec1.html) in 2019 doesn't refer to anything like that. Instead, it mentions funding by [NLnet Foundation's Search and Discovery grant program](https://nlnet.nl/discovery/), which to me seems far outside of blockchain circles (the foundation supports [serious infrastructure stuff](https://nlnet.nl/project/current.html), including [Wireguard integration in the Linux kernel](https://nlnet.nl/project/WireGuard-upscale/)).

Reading the text, I got the slight impression that it wasn't written with neutral judgment, but with blockchain-skepticism in mind. If that's right, I'd like to ask you to reconsider related aspects; if that's not right, I sincerely apologize and retract the suggestion! :)

[No rush!]

**Comment by nweaver** [comment]
I'll argue DNS is distributed not decentralized.  The key with decentralized systems is a lack of a central authority set (of course, most "decentralized" systems aren't, thus the scare quotes, but no quotes is just fine IMO.  GNS for example isn't decentralized if you want to rely on .alt for global consistency of a human namespace). 

DNS has explicit central authority sets (., TLDs etc) which act to delegated authority over subpieces to a distributed set of systems.

Its a subtle but critical distinction:  Distributed systems are more vulnerable to censorship, but tend to be much easier to be consistent, lack issues with sybils, tend to have much better failure modes, are easier to debug (e.g. in DNS if your name fails to resolve you know who to blaim, in GNS its 'the hashtable's fault', and GNS for example can't do provable denial of existence).

And yes, I'm a "Blockchain skeptic", but I'm trying to write with as much neutrality as possible.

**Comment by peterthomassen** [comment]
> DNS has explicit central authority sets (., TLDs etc) which act to delegated authority over subpieces to a distributed set of systems.

No, this is not a property of the Domain Name System, as little as it is a property of the Gnu Name System.

The central authority is due to the habit of shared DNS configuration. As far as the name systems are concerned, both are distributed, and centralization is a configuration matter (not a system design matter).

That's why I'd argue that claims that GNS is more decentralized than DNS are unfounded, and actually out of scope as far as the *systems* are concerned.

It's true that GNS tries to make decentralization easy, but that doesn't make it unfit for a centralized configuration shared by most users.
